File: leetcode-js/060-permutation-sequence.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getPermutation = function(n, k) {
    const nums = [];
    const factorials = [1];
    let sum = 1;

    for (let i = 1; i <= n; i += 1) {
        nums.push(i);
        sum *= i;
        factorials[i] = sum;
    }

    k -= 1;

    let result = '';

    for (let i = 1; i <= n; i += 1) {
        const index = Math.floor(k / factorials[n-i]);
        result += nums[index];
        nums.splice(index, 1);
        k -= index * factorials[n-i];
    }

    return result;
};

File: leetcode-java/060-permutation-sequence.java
public class Solution {
    public String getPermutation(int n, int k) {
        int[] factorials = new int[n + 1];
        factorials[0] = 1;

        for (int i = 1; i < n; i++) {
            factorials[i] = factorials[i - 1] * i;
        }

        List<Integer> nums = new LinkedList<>();

        for (int i = 1; i <= n; i++) {
            nums.add(i);
        }

        k -= 1;

        StringBuilder sb = new StringBuilder();

        for (int i = n - 1; i >= 0; i--) {
            int index = k / factorials[i];
            sb.append(nums.remove(index));
            k %= factorials[i];
        }

        return sb.toString();
    }
}
